---
title: "scRNAseq_PIPELINE PART II"
subtitle: "Murine edit: Analysis of a single cell type"

date: today
date-format: "MM/DD/YYYY"

toc: true

format: 
  html:
    theme: cosmo
    fontawesome: true
    css: styles.css
    embed-resources: true

editor: source

execute: 
  warning: false
  echo: false
  freeze: auto
---

## INTRO: Downstream nalysis of a single cell type

### Purpose

This is represents **Part II** of the **murine** data analysis pipeline and is applicable to multiple samples (typical murne experiment settings). It allows users to analyze a specific cell type of interest from the merged Seurat object generated in **Part I**. This workflow is based upon the [Seurat PBMC 3k tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial) and has been adapted to include downstream cell-specific visualizations. 

For this workflow, you will load the **.rds** file of your merged Seurat object created in the previous workflow.  You will isolate a cell type of interest either by annotaion (e.g. _MouseRNAseq.main_).  You will then perform reclustering, differential gene expression (DEG), generate gene-level visualizations, and create volcano plots based on the DEG results.

## STEP 1: Setup and load packages

```{r}
#| label: Setup
#| output: false
#| message: false

#Load Packages
library(Seurat)
library(RColorBrewer)
library(ggpubr)
library(patchwork)
library(SingleCellExperiment)
library(sctransform)
library(SeuratObject)
library(scater)
library(scran)
library(Matrix)
library(reshape2)
library(ggforce)
library(scRepertoire)
library(SingleR)
library(celldex)
library(glmGamPoi)
library(clustree)
library(dittoSeq)
library(future)
library(scales)
library(grid)
library(gridBase)
library(tidyverse)
library(MAST)
library(EnhancedVolcano)
library(ggridges)
library(viridis)

options(Seurat.object.assay.calcn = TRUE)

cbf_3 <- dittoColors(reps = 10000, get.names=FALSE) ## generates a vector of color-blind friendly colors that we can use for visualizations

```

### Load data files for analysis

```{r}
#| label: Load_mergedSO
#| output: false

merged <- readRDS("Trial dataset/mergedSO_2026-02-02.rds") #direct to your merged SO file saved in template 007.

#Edit colors for your groups (Ref here:https://derekogle.com/NCGraphing/resources/colors). The treatment name needs to remain consistent from Part I.
treatment_colors <- c(
  "G4"     = "red3",
  "G2"     = "darkseagreen4",  
  "G3"     = "steelblue2", 
  "G1"     = "gray68"  
)

```

## STEP 2. View Annotations and Isolate celltype of interest

The first chunk below is a repeat celltype annotation step completed in part I; if you already have the visual handy, you can skip this chunk. Next, once you know which cells you would like to keep, subset them below. Choose either step A. (_MouseRNAseq.main_) or step B. (_ImmGen.main_) below to isolate cell type of interest. A few examples pre-identifying cell types of interest are listed (naming can be different; please reference annotation cell name from the legend). To switch cell type of interest, place a # in front of the line you want to censor and remove it from the line (cell type) of interest. Below is an example subsetting Macrophages as identified by Immgen. At the end of this template, you will get a reclusted subset, identify subset clusters, and run treatment comparison analysis.

```{r}
#| label: Cell_annotations
#| fig-cap: 'SingleR cell type UMAPs'

# View annotated cells by treatment group
MouseRNAseq <- DimPlot(merged,
        group.by='MouseRNAseq.main',
        split.by = 'treatment',
        ncol = 2,
        reduction='umap',
        cols=cbf_3) +
labs(title = "MouseRNAseq")

MouseRNAseq

Immgen <- DimPlot(merged,
        group.by='Immgen.main',
        split.by = 'treatment',
        ncol = 2,
        reduction='umap',
        cols=cbf_3) +
labs(title = "Immgen")

Immgen

MouseRNAseqCELLS <- dittoBarPlot(merged,
  var      = "MouseRNAseq.main",   # what fills the stacks (cell types)
  group.by = "treatment",          # one bar per treatment
  scale    = "count",              # or "percent"
  y.breaks = c(2500, 5000, 7500, 10000),
  max = 10000,
  retain.factor.levels = TRUE
) +
  labs(x = NULL, y = "Number of cells") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

ImmGenCELLS <- dittoBarPlot(merged,
  var      = "Immgen.main",   # what fills the stacks (cell types)
  group.by = "treatment",          # one bar per treatment
  scale    = "count",              # or "percent"
  y.breaks = c(2500, 5000, 7500, 10000),
  max = 10000,
  retain.factor.levels = TRUE
) +
  labs(x = NULL, y = "Number of cells") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

(MouseRNAseqCELLS | ImmGenCELLS) &
  theme(legend.position = "right",
        legend.key.size = unit(0.4, "cm"),
        legend.text  = element_text(size = 8),
        legend.title = element_text(size = 9)
  )

```

### A. Isolate cells by annotaion (ex. MouseRNAseq.main: Macrophages)

```{r}
#| label: Isolate cells by MouseRNAseq annotaion 
#| fig-cap: 'Subsetted cells_MouseRNAseq'
#| eval: FALSE

# Use SCT as the working assay for this part
#DefaultAssay(merged) <- "SCT"

#Once you know which cells you would like to keep, subset them below. A few examples pre-identifying cell types of interest are listed (naming can be different; please reference annotation cell name from the legend). To switch cell type of interest, place a # in front of the line you want to censor and remove it from the line of interest. Below is an example of looking only at Macrophages.

# Name the cells of interest here:
#Idents(merged) <- "MouseRNAseq.main" #identify column to use
#T_Cells <- WhichCells(merged, idents = c("T cells"))  # <-- your label(s)
#Macrophages <- WhichCells(merged, idents = c("Macrophages"))
#B_Cells <- WhichCells(merged, idents = c("B cells"))
#NK_Cells <- WhichCells(merged, idents = c("NK cells"))
#DCs <- WhichCells(merged, idents = c("Dendritic cells"))
#Granulocytes <- WhichCells(merged, idents = c("Granulocytes"))

# Create new SO for those cells only:
#T_CellsSO <- subset(merged, cells = T_Cells) #creates a new SO of your cells 
#MacrophagesSO <- subset(merged, cells = Macrophages)
#B_CellsSO <- subset(merged, cells = B_Cells)
#NK_CellsSO <- subset(merged, cells = NK_Cells)
#DCsSO <- subset(merged, cells = DCs)
#NeutrophilsSO <- subset(merged, cells = Granulocytes)

# View cells of interest against all cells
#selected_cells <- UMAPPlot(
#  merged,
#  reduction = "umap",
#  cells.highlight = list("Macrophages" = Macrophages),  # legend name,
#  cols = c("grey85", "red"),
#  shuffle = TRUE,
#  split.by = 'treatment',
#  ncol = 2, 
#  pt.size = 0.001,
# sizes.highlight = 0.001
#)

#table(merged$treatment[Macrophages])

#UMAP (Pull out cell type of interest separately)
#selected_cluster <- UMAPPlot(MacrophagesSO, 
#               group.by = "MouseRNAseq.main", 
#               split.by = "treatment",
#               ncol = 2,
#               cols = "steelblue")

#selected_cells
#selected_cluster

```

### B. Isolate cells by annotaion (ex. ImmGen: Macrophages)

```{r}
#| label: Isolate cells by ImmGen annotaion
#| fig-cap: 'Subsetted cells_ImmGen'
#| message: FALSE


# Use SCT as the working assay for this part
DefaultAssay(merged) <- "SCT"

# This is an alternate visualization for part A (ImmGen instead of MouseRNAseq)
Idents(merged) <- "Immgen.main" #identify column to use

Macrophages_IG <- WhichCells(merged, idents = c("Macrophages"))

MacrophagesSO_IG <- subset(merged, cells = Macrophages_IG) #if changed, edit name of subsetted SO throughout template using find/replace function

selected_cells_1 <- UMAPPlot(
  merged,
  reduction = "umap",
  cells.highlight = list("Macrophages" = Macrophages_IG),  # legend name,
  cols = c("grey85", "red1"),
  shuffle = TRUE,
  split.by = 'treatment',
  ncol = 2, 
  pt.size = 0.001,
  sizes.highlight = 0.001
)

table(merged$treatment[Macrophages_IG])

selected_cells_1

```

## STEP 3. Recluster Macrophages (e.g. for ImmGen: macrophages)

This step will recluster the subsetted macrophages from Immgen annotation for downstream breakdown of Macrophage cells and visualizations. Select dims accordingly from elbow plot. You can use refer to [Seurat: Determine the ‘dimensionality’ of the dataset](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) for more information about selecting dims. Some pipelines prefer to re-run SCTransform on subsetted cells, though not neccessary. You can run or mute the **Rerun SCTransform** chunk. It may be useful to run with and without re-running SCTransform, and select the pipeline that is supported by complementary experimental data (i.e. results from flow cytometry, _in vitro_ experiments, or cytokine arrays, etc.). 

```{r}
#| label: Rerun SCTransform 
#| eval: FALSE

#This step is optional, it may overcorrect your samples and lose biological significance.  It's recommended to run both if you'd prefer but skipping this step is okay.

#options(future.globals.maxSize = 4 * 1024^3)  # 4 GB

#DefaultAssay(MacrophagesSO_IG) <- "RNA" 

#MacrophagesSO_IG <- SCTransform(MacrophagesSO_IG,
                                #residual.type = "pearson",
                                #do.correct.umi = TRUE,
                                #return.only.var.genes = TRUE,
                                #conserve.memory = TRUE,
                                #verbose = FALSE)
```

```{r}
#| label: Dimentionality_reduction
#| fig-cap: 'Elbow plot'
#| message: FALSE

#This step will recluster the subsetted macrophages SO from MouseRNAseq annotation for downstream breakdown of Macrophage cells and visualizations. There is no visual output.

options(future.globals.maxSize = 4 * 1024^3)  #do not rerun this line if you've completed SCTransform above.

DefaultAssay(MacrophagesSO_IG) <- "SCT" #Edit name of SO only for all lines

set.seed(42)
MacrophagesSO_IG <- RunPCA(MacrophagesSO_IG, npcs = 30, seed.use = 42)  

ElbowPlot(MacrophagesSO_IG, ndims = 30) # use this to select dims

```

```{r}
#| label: Reclustering_across_resolutions
#| message: FALSE
#| include: FALSE

MacrophagesSO_IG <- FindNeighbors(MacrophagesSO_IG, dims = 1:15)          
MacrophagesSO_IG <- FindClusters(MacrophagesSO_IG, resolution = seq(0.2,1.2,0.2))
MacrophagesSO_IG <- RunUMAP(MacrophagesSO_IG, reduction = "pca", dims = 1:15, seed.use = 42)
#MacrophagesSO_IG <- RunTSNE(MacrophagesSO_IG, reduction = "pca", dim.embed = 2, dims = 1:20, seed.use = 1) #optional if you want to see TSNE

SaveSeuratRds(MacrophagesSO_IG, file = paste0("Output/MacrophagesSO_IG", Sys.Date(), ".rds")) #confirm output folder or edit directory 

rm(merged)

```

```{r}
#| label: Reclustered_UMAPS
#| fig-cap: 'Reclustered UMAPS'

## Plot by treatment only
dimplot1 <- DimPlot(MacrophagesSO_IG,
        group.by ='treatment',
        #split.by = 'treatment', 
        #ncol = 2,
        reduction ='umap',
        cols = treatment_colors,
        label = FALSE) +
 labs(title = "By treatment")

dimplot1

## Plot by resolution; change the resolution below as needed
dimplot2 <- DimPlot(MacrophagesSO_IG,
        group.by='SCT_snn_res.0.4',
        split.by = 'treatment',            #if you'd like to see a combined UMAP add a # in front of this line
        ncol = 2,
        reduction='umap',
        cols=cbf_3,
        label=FALSE, 
        pt.size = 0.01) +
 labs(title = "UMAP by res 0.4")

dimplot2

```

## STEP 4. Identify clusters by cell type 

### A. Visualize major markers present per cluster
Visualizations here can help you identify subset clusters (e.g. M1 vs M2)
```{r}
#| label: Cluster_visualizations_and_identification
#| message: FALSE
#| fig-cap: 'Cluster identification of subsetted cells'

Genes_to_plot <- c('Nos2', 'Cxcl9', 'Cd38', 'Arg1', 'Mrc1', 'Cd163', 'Tnf','Il6', 'Vegfa', 'Ly6c1', 'H2-Ab1') #small list for ridge, violin and dot plots below.  Edit and test genes here or plug in under features.  

# 1.CREATE TABLE OF COUNTS PER CLUSTER FOR EACH TX
counts_clusters <- with(MacrophagesSO_IG@meta.data, table(SCT_snn_res.0.4, treatment)) #Get quick counts for clusters

clusters_df <- as.data.frame(counts_clusters) %>%
  rename(cluster = SCT_snn_res.0.4, treatment = treatment, n = Freq) %>%
  group_by(treatment) %>%
  mutate(percent = 100 * n / sum(n)) %>%
  ungroup()

ClusterA <- ggplot(clusters_df, aes(x = treatment, y = percent, fill = cluster)) +
  geom_col(width = 0.9, color = "black", linewidth = 0.1) +
  scale_fill_manual(values = cbf_3) +         
  ylab("Percent of cells") +
  xlab("Treatment") +
  theme_classic()

clusters_df_B <- as.data.frame(counts_clusters) %>%
  rename(cluster = SCT_snn_res.0.4, treatment = treatment, n = Freq) %>%
  group_by(cluster) %>%
  mutate(percent = 100 * n / sum(n)) %>%
  ungroup()

ClusterB <- ggplot(clusters_df_B, aes(x = cluster, y = percent, fill = treatment)) +
  geom_col(width = 0.9) +
  scale_fill_manual(values = treatment_colors) +
  ylab("Percent of cells") +
  xlab("SCT_snn_res.0.4") +
  theme_classic()

(ClusterA | ClusterB) &
  theme(legend.position = "right")

# 2.CREATE STACKED VIOLINPLOT OF GENE EXPRESSION PER CLUSTER
Cluster_genes <- VlnPlot(
  MacrophagesSO_IG,
  features = Genes_to_plot,
  group.by = "SCT_snn_res.0.4",
  stack = TRUE,
  flip = TRUE
) + theme(legend.position = "none")

Cluster_genes

# 3.CREATE A RIDGEPLOT TO VIEW YOUR GENES OF INTEREST PER CLUSTER
RidgePlot(MacrophagesSO_IG, 
          features = c("Mrc1", "Nos2", "Cxcl9", "Arg1"), #or you can plug in 'Genes_to_plot"
          same.y.lims = FALSE, 
          sort = FALSE,                                  #highest expressing group will be in the front if TRUE; note cluster ID not color
          group.by = "SCT_snn_res.0.4", 
          layer = "data", 
          ncol = 2, 
          cols = cbf_3, 
          log = TRUE) 

# 4.CREATE A VIOLINPLOT PER GENE TO VIEW YOUR GENES OF INTEREST BY CLUSTERS
VlnPlot(MacrophagesSO_IG, 
        features = c("Vegfa", "Nos2", "Cxcl9", "Arg1"), 
        group.by = "SCT_snn_res.0.4", 
        same.y.lims = FALSE, 
        layer = "data", 
        pt.size = 0,
        ncol = 2, 
        cols = cbf_3) 

# 5. CREATE A DOTPLOT PER GENE TO VIEW YOUR GENES OF INTEREST BY CLUSTERS
DotPlot(MacrophagesSO_IG, 
        features = Genes_to_plot, 
        assay = 'SCT', 
        cols = c("darkblue", "red2"), 
        group.by = "SCT_snn_res.0.4", 
        scale.by = 'size') +
  coord_flip() +
  theme_light() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), # try 60 or 90 if needed
    axis.title.x = element_text(margin = margin(t = 8))           # change margin to add space   
  )

# 6. CREATE HEATMAP OF GENE EXPRESSION BY CLUSTERS
  
MacrophagesSO_IG <- PrepSCTFindMarkers(MacrophagesSO_IG)

Idents(MacrophagesSO_IG) = "SCT_snn_res.0.4" #edit depending on which resolution you'd like to explore 
Counts_res <- table(Idents(MacrophagesSO_IG)) #table of counts per cluster

#DE by resolution clusters
#This calculates differentially expressed genes for clusters at resolution x.
Markers_res <- FindAllMarkers(
  MacrophagesSO_IG, 
  only.pos = TRUE,
  min.pct = 0.10,               #must be found in this percent of cells
  logfc.threshold = 0.10        #Default is 0.1; increasing logfc.threshold can miss weaker signals 
) 

#Create heatmap for clusters
top_res <- Markers_res %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  slice_head(n = 5) %>%
  ungroup() 

DoHeatmap(
  MacrophagesSO_IG,
  features = top_res$gene,
  group.by = "SCT_snn_res.0.4",
  size = 4,
  angle = 0
) +
  theme(axis.text.y = element_text(size = 4))

```

### B. Set cluster identities
```{r}
#| label: Cluster_id_UMAPs
#| fig-cap: 'Cluster identities annotation'

subset_cluster_ids <- c("Cluster0", "Cluster1", "Classic M2", "Arg1+ M2", "Cluster4", "Cluster5", "Classic M1", "Cluster7", "TNF+ M1", "Cluster9", "IL6+ M1", "inflammatory TAM")                     #Name clusters per your investigative results from above

names(subset_cluster_ids) <- levels(MacrophagesSO_IG)

MacrophagesSO_IG <- RenameIdents(MacrophagesSO_IG, subset_cluster_ids)

DimPlot(MacrophagesSO_IG, 
        reduction = "umap", 
        label = FALSE,
        cols = cbf_3,
        split.by = 'treatment',
        pt.size = 0.1,
        ncol = 2) +
  theme(legend.position = "right")

# Subset Classic M1 and M2 for plots
m1m2 <- subset(MacrophagesSO_IG, idents = c("Classic M1", "Classic M2"))

# Bar plot for counts per treatment 
m1m2_counts <- ggplot(m1m2@meta.data, aes(x = treatment, fill = Idents(m1m2))) +
  geom_bar(color = "black", linewidth = 0.1) +
  ylab("Counts") + xlab(NULL) +
  theme_classic()

# Bar plot for frequency per treatment
m1m2_props <- ggplot(m1m2@meta.data, aes(x = treatment, fill = Idents(m1m2))) +
  geom_bar(position = "fill", color = "black", linewidth = 0.1) +
  scale_y_continuous(labels = scales::percent_format()) +
  ylab("Percent of total") + xlab(NULL) +
  theme_classic()

(m1m2_counts | m1m2_props) &
  theme(legend.position = "right",
        legend.key.size = unit(0.4, "cm"),
        legend.text  = element_text(size = 8),
        legend.title = element_text(size = 9)
  )

```

### C. Visualizations by treatment
```{r}
#| label: Subsetted_cells_by_treatment
#| message: FALSE
#| fig-cap: 'Subsetted cells by treatment'

# 1.CREATE A RIDGEPLOT TO VIEW YOUR GENES OF INTEREST PER TX
RidgePlot(MacrophagesSO_IG, 
          features = c("Mrc1", "Nos2", "Cxcl9", "Arg1"), #or you can plug in 'Genes_to_plot"
          same.y.lims = FALSE, 
          sort = FALSE,                                  #highest expressing group will be in the front.
          group.by = "treatment", 
          layer = "data", 
          ncol = 2, 
          cols = cbf_3, 
          log = TRUE) 

# 2.CREATE A VIOLINPLOT PER GENE TO VIEW YOUR GENES OF INTEREST BY TX
VlnPlot(MacrophagesSO_IG, 
        features = c("Mrc1", "Nos2", "Cxcl9", "Arg1"), 
        group.by = "treatment", 
        same.y.lims = FALSE,
        layer = "data", 
        pt.size = 0,                                       #edit if you want to show all cells (busy)
        ncol = 2, 
        cols = cbf_3) 

# 3. CREATE A DOTPLOT PER GENE TO VIEW YOUR GENES OF INTEREST BY TX
DotPlot(MacrophagesSO_IG, 
        features = c('Nos2', 'Cxcl9', 'Cd38', 'Arg1', 'Mrc1', 'Cd163', 'Tnf','Il6', 'Vegfa', 'Ly6c1'), #or use Genes_to_plot created earlier
        assay = 'SCT', 
        cols = c("darkblue", "red2"), 
        group.by = "treatment",
        scale = FALSE,
        scale.by = 'size') +
  coord_flip() +
  theme_light() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), # try 60 or 90 if needed
    axis.title.x = element_text(margin = margin(t = 8))           # change margin to add space   
  )

# 4. CREATE single gene UMAPs BY TX
singleUMAP <- FeaturePlot(
  MacrophagesSO_IG,
  features = c('Nos2', 'Cxcl9', 'Cd38', 'Arg1', 'Mrc1', 'Cd163', 'Tnf'), # edit as needed; will help in identifying which cell types are expressing what gene
  cols = c("grey90","navyblue"),
  pt.size = 0.1,
  split.by = "treatment",
  keep.scale = "all",
  by.col = FALSE,
  ncol = 2
)

singleUMAPfigure <- wrap_plots(singleUMAP) +
  plot_layout(guides = "collect") &   # <— collects into one legend
  guides(color = guide_colorbar(
    title = "Expression",
    direction = "vertical",
    title.position = "top",
    barwidth = unit(0.35, "cm"),
    barheight = unit(2, "cm")
  ))&
  theme(
    legend.position = "right",
    legend.box = "vertical",
    legend.title = element_text(size = 11),
    legend.text  = element_text(size = 10),
    axis.title   = element_text(size = 10),
    axis.text    = element_text(size = 10)
  )

print(singleUMAPfigure)
invisible(singleUMAP) 

#Get quick counts by genes
countsbygene <- t(rowsum(t((as.matrix(GetAssayData(MacrophagesSO_IG, layer="data")[Genes_to_plot, ]) > 0) + 0),
                   MacrophagesSO_IG$treatment))

```

### D. Differential expression analysis by treatment group

#### I. Calculate DE for subsetted cells
```{r}
#| label: DEG
#| message: FALSE

# How many cells per group?
Idents(MacrophagesSO_IG) <- "treatment"

#Change contrasts as needed
DEG_list <- FindMarkers(MacrophagesSO_IG, #This gives gene expression markers for all treatments.
              ident.1 = "G4",
              ident.2 = "G1",
              test.use = "wilcox",
              #latent.vars = c("nCount_RNA","percent.mt"), # no covariates
              min.pct = 0.2, # Drops ultra-rare genes that are mostly zeros; will pass if that gene is ≥25% in any group
              min.diff.pct = 0.02, # Drops genes with small absolute difference between groups (i.e. 10% vs 12% will get dropped but will keep if 10% vs 15% in G1 vs G2; keep at 0.02-0.05)
              min.cells.feature = 3,  # absolute gate (min in either group)
              logfc.threshold = 0,
              #verbose = FALSE,
              slot = "data")  # test all genes

# Save the original DEG table (records)
write.csv(DEG_list, "DEG_G4_vs_G1.csv", row.names = TRUE)

# Create a version for volcano plots below
DEG_volcano <- DEG_list |>
  tibble::rownames_to_column(var = "Gene") |>
  dplyr::select(-c(4, 5))  # drop columns D-E by position

```

#### II. Prepare file for volcano

```{r}
#| label: Prepare data for volcano
#| message: FALSE

# Ensure you have a data frame with 'log2FC', 'pvalue', and 'gene' columns
# Exclude ribosomal genes, gm-genes, and mitochondrial genes.
volcano_clean <- subset(
  DEG_volcano,
  !grepl("^Gm\\d+$", Gene, perl = TRUE) &                # predicted gene models like Gm12345
  !grepl("Rik$",    Gene) &                 # RIKEN cDNA names, e.g. 1700019G24Rik
  !grepl("^(Rps|Rpl)", Gene, ignore.case=TRUE) &  # ribosomal
  !grepl("^mt-",      Gene, ignore.case=TRUE)     # mitochondrial
)

# assign colnames 
volcano_clean <- volcano_clean |>
  dplyr::rename(
    log2FoldChange = "avg_log2FC",     
    pvalue = "p_val",                   
    padj = "p_val_adj"                   
    )
    
```

#### III. Create Volcano Plot

```{r}
#| label: Volcano plot
#| message: FALSE

MACG4_G1 <- EnhancedVolcano(volcano_clean,
    lab = volcano_clean$Gene,
    selectLab = c('Ly6c2', 'Arg1', 'Cd163', 'Tnf', 'Mrc1', 'Cd38', 'Cxcl2', 'Nos2', 'Cxcl9', 'Fcgr4', 'Ccr5', 'H2-Aa',  'Ccl5', 'Cxcl10', 'Cd274', 'H2-DMb2', 'Ly6a'),
    x = 'log2FoldChange',
    y = 'pvalue',
    xlim =c(-2.5, 6.5),
    ylim = c(0,325),
    borderWidth = 0.5,
    title = 'G4 vs G1',
    xlab = bquote(~Log[2]~ 'fold change'),
    ylab = bquote(~-Log[10]~ 'p-value'),
    pCutoff = 0.05,
    FCcutoff = 0.263,
    cutoffLineType = 'twodash',
    cutoffLineWidth = 0.8,
    pointSize = 0.05,
    labSize = 3.0,
    labCol = 'black',
    #labFace = 'bold',
    boxedLabels = FALSE,
    parseLabels = FALSE,
    col = c('gray40', 'green4', 'blue', 'red3'),
    colAlpha = 1/2,
    legendLabels=c('NS','Log (base 2) FC','p-value',
      'p-value & Log (base 2) FC'),
    legendPosition = 'right',
    legendLabSize = 8,
    legendIconSize = 2.0,
    drawConnectors = TRUE,
    max.overlaps = 20,
    lengthConnectors = unit(0.01, "npc"),
    widthConnectors = 0.5,
    colConnectors = 'black',
    #gridlines.major = FALSE,
    gridlines.minor = FALSE,
)

plot(MACG4_G1)

```

